gateway/src/apicast/oauth/oidc.lua

```diff
@@ -92,64 +95,99 @@ end
 -- Parses the token - in this case we assume it's a JWT token
 -- Here we can extract authenticated user's claims or other information returned in the access_token
 -- or id_token by RH SSO
-local function parse_and_verify_token(self, jwt_token)
+local function parse_and_verify_token(self, namespace, jwt_token)
   local cache = self.cache
 
   if not cache then
     return nil, 'not initialized'
   end
-  local cache_key = format('%s:%s', self.service.id, jwt_token)
+  local cache_key = format('%s:%s', namespace, jwt_token)
+
+  local jwt = self:parse(jwt_token, cache_key)
+
+  if jwt.verified then
+    return jwt
+  end
+
+  local _, err = self:verify(jwt, cache_key)
+
+  return jwt, err
+end
+
+function _M:parse_and_verify(access_token, cache_key)
+  local jwt_obj, err = parse_and_verify_token(self, assert(cache_key, 'missing cache key'), access_token)
+
+  if err then
+    if ngx.config.debug then
+      ngx.log(ngx.DEBUG, 'JWT object: ', require('inspect')(jwt_obj), ' err: ', err, ' reason: ', jwt_obj.reason)
+    end
+    return nil, jwt_obj and jwt_obj.reason or err
+  end
+
+  return jwt_obj
+end
+
+local jwt_mt = {
+  __tostring = function(jwt)
+    return jwt.token
+  end
+}
+
+local function load_jwt(token)
+  local jwt = JWT:load_jwt(tostring(token))
 
-  local jwt_obj = cache:get(cache_key)
+  jwt.token = token
 
-  if jwt_obj then
+  return setmetatable(jwt, jwt_mt)
+end
+
+function _M:parse(jwt, cache_key)
+  local cached = self.cache:get(cache_key or jwt)
```

Comment:
Why `cache_key or jwt` ? Shouldn't it be just `cache_key` ?

Reply:
Because `cache_key` can be `nil`. So the token itself is its cache key.